### PR TITLE
Pipelines on settings page

### PR DESCRIPTION
**Pull request by SamardakOleh**

### CODE REVIEW

File: api/funnel/index.test.js [hunk 1]
```diff
@@ -23,7 +23,7 @@ describe("Funnel", function() {
       .set("Authorization", cred.token)
       .send({ domain: cred.domainId, name: "Sales Funnel" });
     expect(status).toBe(200);
-    expect(typeof body).toBe("string");
+    expect(typeof body).toBe("object");
```
Comment (valentyntu) [inline api/funnel/index.test.js]: you have to check if object has proper data in fields, not just its type

File: api/stage/index.test.js [hunk 2]
```diff
@@ -30,7 +31,7 @@ describe("Stage", () => {
       });
 
     expect(status).toBe(200);
-    expect(typeof body).toBe("string");
+    expect(typeof body).toBe("object");
```
Comment (valentyntu) [inline api/stage/index.test.js]: you have to check if object has proper data in fields, not just its type

File: client/src/modules/settings/Funnels/Funnel/AddStageModal/AddStageModal.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+  }
+
+  private onSave = () => {
+    this.props.onSave(this.state.name, this.props.stage ? this.props.stage._id : undefined);
+  }
+
+  private onNameChange = (e: any) => {
+    this.setState({
+      ...this.state,
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnel/AddStageModal/AddStageModal.tsx]: why do you copy state's fields here? Name is the only field and it is going to be changed below

File: client/src/modules/settings/Funnels/Funnel/Funnel.tsx [hunk 4]
```diff
@@ -66,60 +76,92 @@ class Funnel extends React.Component<Props, State> {
...
+        isEditNameModalOpen: true,
         name: this.props.funnel.name,
       })
     }
   }
 
+  private openAddStageModal = () => {
+    this.setState({
+      isAddStageModalOpen: true,
+    })
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnel/Funnel.tsx]: semicolon?

File: client/src/modules/settings/Funnels/Funnel/Funnel.tsx [hunk 5]
```diff
@@ -66,60 +76,92 @@ class Funnel extends React.Component<Props, State> {
...
       })
     }
   }
 
-  private cancel = () => {
+  private onEditNameCancel = () => {
+    this.setState({
+      ...this.state,
+      isEditNameModalOpen: false,
+    })
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnel/Funnel.tsx]: semicolon

File: client/src/modules/settings/Funnels/Funnel/Funnel.tsx [hunk 6]
```diff
@@ -66,60 +76,92 @@ class Funnel extends React.Component<Props, State> {
...
+  }
+
+  private onAddNewStageSave = (name: string) => {
+    let stage: any = {
+      funnel: this.props.funnel ? this.props.funnel._id : null,
+      name,
+      order: this.props.stages ? this.props.stages.length+1 : null,
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnel/Funnel.tsx]: "length+1" should be like "length + 1"

File: client/src/modules/settings/Funnels/Funnel/Funnel.tsx [hunk 7]
```diff
@@ -66,60 +76,92 @@ class Funnel extends React.Component<Props, State> {
...
-      isModalOpen: false,
+      isAddStageModalOpen: false,
     })
   }
 }
 
-const mapStateToProps = (state: any) => ({})
+const mapStateToProps = (state: any) => ({
+  stages: state.settings.stages,
+})
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnel/Funnel.tsx]: semicolon

File: client/src/modules/settings/Funnels/Funnel/Stages/Stages.tsx [hunk 8]
```diff
@@ -1,10 +1,86 @@
...
+      </div>
     )
   }
+
+  private onEditStageSave = (name: string, id: string) => {
+    this.props.updateStageName(id, name);
+    this.setState({
+      isModalOpen: false,
+    })
+  }
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnel/Stages/Stages.tsx]: semicolon

File: client/src/modules/settings/Funnels/Funnel/Stages/Stages.tsx [hunk 9]
```diff
@@ -1,10 +1,86 @@
...
+        />
+      </div>
     )
   }
+
+  private onEditStageSave = (name: string, id: string) => {
+    this.props.updateStageName(id, name);
+    this.setState({
+      isModalOpen: false,
+    })
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnel/Stages/Stages.tsx]: semicolon

File: client/src/modules/settings/Funnels/Funnels.tsx [hunk 10]
```diff
@@ -1,34 +1,105 @@
...
+            onInputBlur={this.onInputBlur}
+            isInputEmpty={this.state.isInputEmpty}
+            inputValue={this.state.inputValue}
+            onSaveButtonClick={this.onSaveButtonClick}
+          />
+        </div>
+        <div className={styles.tabs}>
+          {this.createTabs(this.props.funnels)}
+        </div>
+        <Funnel funnel={this.props.selectedFunnel}/>
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnels.tsx]: missing white space before "/>"

File: client/src/modules/settings/Funnels/Funnels.tsx [hunk 11]
```diff
@@ -1,34 +1,105 @@
...
 
   public componentWillReceiveProps(nextProps: Props) {
     if (nextProps.domainId && this.props.funnels.length === 0) {
       this.props.loadFunnels();
     }
   }
+
+  public componentWillMount() {
+    this.props.loadFunnels();
+  }
```
Comment (valentyntu) [inline client/src/modules/settings/Funnels/Funnels.tsx]: semicolon

File: api/stage/index.js [hunk 12]
```diff
@@ -38,7 +55,27 @@ router.post("/", function(req, res) {
...
+    .catch(error => {
+      res.status(400).json({ errors: { message: error } });
+    });
+});
+
+// @route   PATCH api/stage/:stageId
+// @desc    Update stage
+// @access  Private
+router.patch("/:stageId", validateStageDomain, (req, res) => {
+  if (isBlank(req.body.name)) {
```
Comment (valentyntu) [inline api/stage/index.js]: won't work if you are patching any other field

File: client/src/modules/settings/Funnels/AddPipelineModal/AddPipelineModal.tsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+                id="pipeline-name"
+                onChange={this.props.onInputChange}
+                onBlur={this.props.onInputBlur} 
+              />
+            </div>
+            <div className={styles.formControl}>
+              <button
+                className={isBlank(this.props.inputValue) ? styles.disabledSaveButton : styles.enabledSaveButton}
+                disabled={isBlank(this.props.inputValue)}
+                onClick={this.props.onSaveButtonClick}
```
Comment (alexanderko) [inline client/src/modules/settings/Funnels/AddPipelineModal/AddPipelineModal.tsx]: use form onSubmit

File: client/src/modules/settings/Funnels/Funnel/AddStageModal/AddStageModal.tsx [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+              <input 
+                className={styles.input} 
+                type="text" 
+                value={this.state.name}
+                onChange={this.onNameChange}
+              />
+              <div className={styles.modalButtons}>
+                <button 
+                  disabled={isBlank(this.state.name)} 
+                  onClick={this.onSave}
```
Comment (alexanderko) [inline client/src/modules/settings/Funnels/Funnel/AddStageModal/AddStageModal.tsx]: use form onSubmit

File: client/src/modules/settings/Funnels/Funnel/Funnel.tsx [hunk 15]
```diff
@@ -66,60 +75,92 @@ class Funnel extends React.Component<Props, State> {
...
-      })
+        isEditNameModalOpen: false,
+      });
     }
   }
 
-  private cancel = () => {
+  private onEditNameCancel = () => {
+    this.setState({
+      ...this.state,
```
Comment (alexanderko) [inline client/src/modules/settings/Funnels/Funnel/Funnel.tsx]: no need to duplicate state here

File: client/src/modules/settings/Funnels/Funnel/Funnel.tsx [hunk 16]
```diff
@@ -66,60 +75,92 @@ class Funnel extends React.Component<Props, State> {
...
+
+  private onAddNewStageSave = (name: string) => {
+    const stage: any = {
+      funnel: this.props.funnel ? this.props.funnel._id : null,
+      name,
+      order: this.props.stages ? this.props.stages.length + 1 : null,
+    };
+    this.props.createStage(stage);
+    this.setState({
+      ...this.state,
```
Comment (alexanderko) [inline client/src/modules/settings/Funnels/Funnel/Funnel.tsx]: no need to duplicate state here

File: client/src/modules/settings/Funnels/Funnel/Stages/StageView/StageView.tsx [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+  public render() {
+    return (
+      <div className={styles.container}>
+        <div className={styles.content}>
+          <span className={styles.stageName}>
+            {this.props.stage.name}
+          </span>
+        </div>
+    </div>
+  );
```
Comment (alexanderko) [inline client/src/modules/settings/Funnels/Funnel/Stages/StageView/StageView.tsx]: one indentation missed

File: client/src/modules/settings/Funnels/Funnels.tsx [hunk 18]
```diff
@@ -1,34 +1,104 @@
...
+    );
   }
 
   public componentWillReceiveProps(nextProps: Props) {
     if (nextProps.domainId && this.props.funnels.length === 0) {
       this.props.loadFunnels();
     }
   }
+
+  public componentWillMount() {
```
Comment (alexanderko) [inline client/src/modules/settings/Funnels/Funnels.tsx]: `componentDidMount`

File: client/src/modules/settings/Funnels/Funnels.tsx [hunk 19]
```diff
@@ -1,34 +1,104 @@
...
+  }
+
+  private onCancelClick = () => {
+    this.setState({ isModalOpen: false });
+  }
+
+  private createTabs = (funnels: FunnelModel[]) => {
+    return funnels.map((funnel: FunnelModel) => (
+      <button
+        key={`${funnel.name}`}
```
Comment (alexanderko) [inline client/src/modules/settings/Funnels/Funnels.tsx]: no need in interpolation

File: client/src/modules/settings/settingActions.ts [hunk 20]
```diff
@@ -1,47 +1,80 @@
...
+
+export const createFunnel = (name: string) => (dispatch: Dispatch) => {
+  axios
+    .post(`/api/funnel`, { name })
+    .then(result => {
+      dispatch({
+        payload: result.data,
+        type: ADD_FUNNEL,
+      });
+    });
```
Comment (alexanderko) [inline client/src/modules/settings/settingActions.ts]: catch request failure

File: validation/stage.js [hunk 21]
```diff
@@ -8,7 +8,7 @@ export const validateStageInput = data => {
   if (isEmpty(data.funnel)) errors.funnel = "Funnel ID cannot be empty";
   if (!isNumber(data.order)) errors.order = "Order must be a number";
   if (!Validator.isLength(data.name, { max: 30 })) {
-    errors.name = "Stage name cannot be more 30 characters";
+    errors.name = "Funnel name cannot be more 30 characters";
```
Comment (alexanderko) [inline validation/stage.js]: Stage?

**Pull request merged**